Field guide modal

srallen opened the pull request:
Package:

Closes #391.

Describe your changes:
Adds the UI for the field guide

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?
- [ ] Is the changelog updated?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?

srallen commented:
My project 1233 has a field guide. The default project loaded by the classifier does not.

srallen commented:
I don't have this setup for the light/dark themes yet. This is an already large PR so I'll note in an issue.

srallen commented:
This also has the issue of the size of the modal changing between the full list of items and selected item. I haven't come up with a better way to have it be responsive and have a max-width yet...

eatyourgreens commented:
Testing this out in VoiceOver, I'm wondering if the item buttons should be links to the individual items? Give it a go and let me know what you think. When I press a button, it's not immediately obvious that I've navigated to the item, although I can continue navigating with VoiceOver and read the new text.

There's an annoying jump in height between an item loading and the linked image loading. Can we show a placeholder for images, or set a default height on that box.

eatyourgreens commented on packages/lib-classifier/src/store/FieldGuideStore.js:
```diff
@@ -92,11 +92,18 @@ const FieldGuideStore = types
-        self.activeItem = index
-        if (fieldGuide.items[index].icon) self.activeMedium = fieldGuide.items[index].icon
+      if (fieldGuide) {
+        const items = fieldGuide.items
```
```suggestion
        const { items } = fieldGuide
```

eatyourgreens commented on packages/lib-classifier/src/store/FieldGuideStore.js:
```diff
@@ -92,11 +92,18 @@ const FieldGuideStore = types
+        const items = fieldGuide.items
+        const index = items.indexOf(item)
+        if (items[index]) {
+          const icon = items[index].icon
```
```suggestion
          const { icon } = items[index]
```

eatyourgreens commented on packages/lib-classifier/src/store/FieldGuideStore.js:
```diff
@@ -92,11 +92,18 @@ const FieldGuideStore = types
+        if (items[index]) {
+          const icon = items[index].icon
+          if (icon) self.activeMedium = icon
+          return self.activeItem = index
```
Why is `self.activeItem` an index, and not the item? I find that confusing.

srallen commented:
The new suggested changes feature is neat.

eatyourgreens commented:
Yeah. I haven't tried it out for changes that are more than a few characters. Big changes might break things.

srallen commented:
I updated to use an anchor instead and I like it a lot better for the hover state than what I came up with for the button.

eatyourgreens commented:
VoiceOver says there's a missing counterpart translation in the link text somewhere.
![Screenshot of the field guide item links, showing a missing translation error for the link text.](https://user-images.githubusercontent.com/59547/55495483-e3b3a480-5634-11e9-8e97-e877b83f7ed6.png)

srallen commented:
@eatyourgreens Oops! That's fixed in the latest commit.

eatyourgreens commented:
I'm also getting `setActiveItem is not a function` when I click those links.

srallen commented:
In addition to the refactor to use Anchor instead of Button for the items list, I've:
- Created an icon component that uses a placeholder fallback consistently between the list items view and the item details view
- Added Grommet's ResponsiveContext component to set the widths for the outer Box container. This fixes it between the item list view and item detail view and also makes it responsive. Note Enzyme does not yet support React's `createContext`, so I cannot test how `FieldGuide` works until Enzyme updates this. See: https://github.com/airbnb/enzyme/issues/1908
- Refactored how I was using Grid. It was just wrong before. I'm still learning!
- Finished the fix for renaming the `setActiveItemIndex` and `activeItemIndex`
- Fix the tests

srallen commented on packages/lib-classifier/src/store/FieldGuideStore.js:
```diff
@@ -92,19 +92,27 @@ const FieldGuideStore = types
     }
 
-    function setActiveItem (index) {
+    function setActiveItemIndex (item) {
```
The action is now using index consistently.

srallen commented:
The focus bug I'll open as a separate issue. It might be a Grommet bug.

eatyourgreens commented:
I've pulled down the rebased branch and the item links aren't working any more. I'll have a look to see if I can figure out why.

eatyourgreens commented on packages/lib-classifier/src/store/FieldGuideStore.js:
```diff
@@ -92,20 +92,14 @@ const FieldGuideStore = types
-        }
-
-        return self.activeItemIndex = undefined
+      if (fieldGuide && fieldGuide.items.length === index + 1 && fieldGuide.items[index]) {
```
`fieldGuide.items.length === index + 1` is only true for the last item can be active at the moment.

eatyourgreens replied:
This should have been caught by a test too.

srallen replied:
I've fixed this and updated the tests too.

eatyourgreens commented on packages/lib-classifier/src/store/FieldGuideStore.spec.js:
```diff
@@ -32,7 +36,7 @@ const fieldGuideWithoutIcon = FieldGuideFactory.build({
 const project = ProjectFactory.build()
 
-describe('Model > FieldGuideStore', function () {
+describe.only('Model > FieldGuideStore', function () {
```
Stray `.only`

srallen replied:
Whoops! That should be removed now.

eatyourgreens commented:
Just waiting on Travis.
